File: HomeAutomation/HomeAutomation/src/HomeAutomation.java
import java.util.ArrayList;
import java.util.List;

public class HomeAutomation implements Subject{
    private List<Observer>observers=new ArrayList<>();
    private final EnvironmentState state=new EnvironmentState();

    public void addObserver(Observer observer){
        observers.add(observer);
    }

    public void removeObserver(Observer observer){
        observers.remove(observer);
    }

    public void updateState(String key,Object value){
        state.updateState(key,value);
        notifyObservers();
    }

    public void notifyObservers(){
        for (Observer observer:observers){
            observer.update(state);
        }
    }
}

File: HomeAutomation/HomeAutomation/Core/HomeAutomation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace HomeAutomation
{
    public class StateManager : IStateSubject
    {
        private readonly List<IStateObserver> _observers = new List<IStateObserver>();
        private EnvironmentalState _currentState = new EnvironmentalState();

        public void RegisterObserver(IStateObserver observer)
        {
            _observers.Add(observer);
        }

        public void RemoveObserver(IStateObserver observer)
        {
            _observers.Remove(observer);
        }

        public void NotifyObservers()
        {
            foreach (var observer in _observers)
            {
                observer.Update(_currentState);
            }
        }

        public void UpdateState(EnvironmentalState newState)
        {
            _currentState = newState;
            NotifyObservers();
        }
    }
}
